### PR TITLE
Sampling-Propagation:Set sampling decision mechanism tag

### PR DESCRIPTION
**Pull request by marcotc**
Follow up from #2266

This PR adds a trace-level tag to propagate the sampling mechanism used to make a sampling decision: `_dd.p.dm`.

The currently defined mechanisms are (many unused):
```ruby
        module Mechanism
          # Used before the tracer receives any rates from agent and there are no rules configured.
          DEFAULT = 0
          # The sampling rate received in the agent's http response.
          AGENT_RATE = 1
          # Auto. Reserved for future use.
          REMOTE_RATE_AUTO = 2
          # Sampling rule or sampling rate based on tracer config.
          TRACE_SAMPLING_RULE = 3
          # User directly sets sampling priority via {Tracing.reject!} or {Tracing.keep!},
          # or by a custom sampler implementation.
          MANUAL = 4
          # Formerly AppSec.
          ASM = 5
          # User-defined target. Reserved for future use.
          REMOTE_RATE_USER = 6
          # Reserved for future use.
          REMOTE_RATE_BY_DATADOG = 7
          # Single Span Sampled.
          SPAN_SAMPLING_RATE = 8
        end
```

The sampling mechanism can either come from an upstream service, if the sampling decision was performed upstream, or added by one of `ddtrace`s sampling classes.

This PR also adds many feature test around different sampling mechanism (e.g. manual sampling, agent-based rates) in order to test for the newly added sampling mechanism tag. This unveiled a few bugs that have been fixed in separated PRs.

## `_dd.p.dm` serialization format

One quirk of the implementation is due to the historical format of `_dd.p.dm` on other tracers: `_dd.p.dm` used to be comprised of two pieces of information, separated by a dash (`-`), e.g. `info1-sampling_mechanism`.

`info1` is now removed but `sampling_mechanism` stays. This means that `_dd.p.dm` will look like this: `-sampling_mechanism`.

This ensures users using old versions of Datadog products that support the `-sampling_mechanism` format can still participate in the sampling propagation chain.

### CODE REVIEW

File: lib/datadog/tracing/sampling/rule_sampler.rb [hunk 1]
```diff
@@ -48,7 +48,7 @@ def initialize(
                                nil
                              else
                                # TODO: Simplify .tags access, as `Tracer#tags` can't be arbitrarily changed anymore
-                               RateByServiceSampler.new(1.0, env: -> { Tracing.send(:tracer).tags[:env] })
+                               RateByServiceSampler.new(1.0, env: -> { Tracing.send(:tracer).tags['env'] })
```
Comment (marcotc) [inline lib/datadog/tracing/sampling/rule_sampler.rb:51]: This is fixed in #2287, but ported here to make CI ✅ .

File: lib/datadog/tracing/sampling/span/sampler.rb [hunk 2]
```diff
@@ -50,11 +50,18 @@ def initialize(rules = [])
...
-              rule.sample!(span_op) != :not_matched
+            # Applies the first matching rule
+            final_decision = nil
+            @rules.each do |rule|
+              decision = rule.sample!(span_op)
+              if decision != :not_matched
+                final_decision = decision
+                break
+              end
             end
```
Comment (marcotc) [inline lib/datadog/tracing/sampling/span/sampler.rb:61]: The final sampling decision wasn't captured in this `any?` loop before, but it's now important to know when to tag the `sampling_mechanism`.

File: lib/ddtrace/transport/trace_formatter.rb [hunk 3]
```diff
@@ -155,6 +156,15 @@ def tag_sample_rate!
         )
       end
 
+      def tag_sampling_mechanism!
+        return unless trace.sampling_mechanism
+
+        root_span.set_tag(
+          Tracing::Metadata::Ext::Distributed::TAG_DECISION_MAKER,
+          "-#{trace.sampling_mechanism}"
```
Comment (marcotc) [inline lib/ddtrace/transport/trace_formatter.rb:164]: Here's the formatting of the "two" part tag: `part1-sampling_mechanism`, where `part1` is temporarily not populated.

**Review by marcotc — COMMENTED** [review]
TODO: make sampling mechanism as a simple tag access.

File: lib/datadog/tracing/sampling/span/sampler.rb [hunk 4]
```diff
@@ -50,9 +50,31 @@ def initialize(rules = [])
...
-            @rules.any? do |rule|
-              rule.sample!(span_op) != :not_matched
+            # Applies the first matching rule
+            @rules.each do |rule|
+              decision = rule.sample!(span_op)
+
+              next if decision == :not_matched # Iterate until we find a matching decision
+
+              if decision == :kept
+                if trace_op.sampled?
```
Comment (marcotc) [inline lib/datadog/tracing/sampling/span/sampler.rb:60]: Move to trace formatter by always setting the decision maker tag in the root span of all trace segments.

Comment (marcotc) [reply lib/datadog/tracing/sampling/span/sampler.rb]: This is done!

**Review by delner — APPROVED** [review]
Thanks for all your hard work on this one! :+1:

**Pull request merged**